File: solution/210CourseScheduleII.py
#BFS
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        inDegree = [ 0 for _ in range(numCourses)]
        neighbor = [ [] for _ in range(numCourses) ]
        ans = []
        queue = []
        for depend in prerequisites:
            inDegree[depend[0]] += 1
            neighbor[depend[1]].append(depend[0])
        for i in range(numCourses):
            if inDegree[i] == 0:
                queue.append(i)
        while queue:
            cur_class = queue.pop()
            ans.append(cur_class)
            for next_class in neighbor[cur_class]:
                inDegree[next_class] -= 1
                if inDegree[next_class] == 0:
                    queue.append(next_class)
        return ans if len(ans) == numCourses else []

#DFS
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        def dfs(cur , visited, isLooped , graph , ans):
            if isLooped[cur] == 1:
                return False
            isLooped[cur] = 1
            for nxt in graph[cur]:
                if visited[nxt] == 0 and not dfs(nxt, visited, isLooped , graph , ans):
                    return False
            isLooped[cur] = 0
            visited[cur] = 1
            ans.append(cur)
            return True

        visited = [0 for _ in range(numCourses)]
        isLooped = [0 for _ in range(numCourses)]
        graph = [[] for _ in range(numCourses)]
        ans = []
        for cur,nxt in prerequisites:
            graph[cur].append(nxt)
        for i in range(numCourses):
            if visited[i] == 0 and not dfs(i, visited, isLooped , graph , ans):
                return []
        print (ans)
        return ans if len(ans) == numCourses else []

File: solution/210CourseScheduleII.cpp
//
// Created by Alaric on 2019-10-10.
//
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        for (const auto& e: prerequisites)
            graph[e[1]].push_back(e[0]);

        // states: 0 = unkonwn, 1 == visiting, 2 = visited
        vector<int> visited(numCourses , 0);
        vector<int> ans;

        for (int i = 0 ; i < numCourses ; ++i)
            if (dfs(i, graph, visited , ans))
                return {};
        reverse(ans.begin() , ans.end());
        return ans;

    }
private:
    bool dfs(int cur , vector<vector<int>>& graph , vector<int>& v , vector<int>& ans){
        //有没有环
        if (v[cur] == 1) return true;
        if (v[cur] == 2) return false;
        v[cur] = 1;
        for (auto e: graph[cur])
            if (dfs(e , graph , v, ans))
                return true;
        v[cur] = 2;
        ans.push_back(cur);
        return false;
    }
};



